go/libraries/doltcore/env/dolt_docs.go

```diff
@@ -15,88 +15,98 @@
-	doltdb.DocDetails{DocPk: doltdb.LicensePk, File: LicenseFile},
-}
+	stgDocTbl, stgDocsFound, err := stgRoot.GetTable(ctx, doltdb.DocTableName)
+	if err != nil {
+		return err
+	}
 
-func LoadDocs(fs filesys.ReadWriteFS) (Docs, error) {
-	docsWithCurrentText := *AllValidDocDetails
-	for i, val := range docsWithCurrentText {
-		path := getDocFile(val.File)
-		exists, isDir := fs.Exists(path)
-		if exists && !isDir {
-			data, err := fs.ReadFile(path)
-			if err != nil {
-				return nil, err
-			}
-			val.NewerText = data
-			docsWithCurrentText[i] = val
-		}
+	_, wrkDocsFound, err := wrkRoot.GetTable(ctx, doltdb.DocTableName)
+	if err != nil {
+		return err
 	}
-	return docsWithCurrentText, nil
-}
 
-func (docs Docs) Save(fs filesys.ReadWriteFS) error {
-	for _, doc := range docs {
-		if !IsValidDoc(doc.DocPk) {
-			continue
-		}
-		filePath := getDocFile(doc.File)
-		if doc.NewerText != nil {
-			err := fs.WriteFile(filePath, doc.NewerText)
-			if err != nil {
-				return err
-			}
-		} else {
-			err := DeleteDoc(fs, doc.DocPk)
-			if err != nil {
-				return err
-			}
+	if wrkDocsFound && !stgDocsFound {
+		newWrkRoot, err := wrkRoot.RemoveTables(ctx, doltdb.DocTableName)
+		if err != nil {
+			return err
 		}
+		_, err = UpdateWorkingRoot(ctx, ddb, rsw, newWrkRoot)
+		return err
 	}
-	return nil
-}
 
-func DeleteDoc(fs filesys.ReadWriteFS, docName string) error {
-	for _, doc := range *AllValidDocDetails {
-		if doc.DocPk == docName {
-			path := getDocFile(doc.File)
-			exists, isDir := fs.Exists(path)
-			if exists && !isDir {
-				return fs.DeleteFile(path)
-			}
+	if stgDocsFound {
+		newWrkRoot, err := wrkRoot.PutTable(ctx, doltdb.DocTableName, stgDocTbl)
+		if err != nil {
+			return err
 		}
+		_, err = UpdateWorkingRoot(ctx, ddb, rsw, newWrkRoot)
+		return err
 	}
 	return nil
 }
 
-func IsValidDoc(docName string) bool {
-	for _, doc := range *AllValidDocDetails {
-		if doc.DocPk == docName {
-			return true
-		}
+// UpdateFSDocsFromRootDocs updates the provided docs from the root value, and then saves them to the filesystem.
+// If docs == nil, all valid docs will be retrieved and written.
+func UpdateFSDocsFromRootDocs(ctx context.Context, root *doltdb.RootValue, docs doltdocs.Docs, FS filesys.Filesys) error {
+	docs, err := doltdocs.GetDocsWithTextFromRoot(ctx, root, docs)
+	if err != nil {
+		return nil
 	}
-	return false
+	return docs.Save(FS)
 }
 
-func hasDocFile(fs filesys.ReadWriteFS, file string) bool {
-	exists, isDir := fs.Exists(getDocFile(file))
-	return exists && !isDir
+// UpdateRootWithDocs takes in a root value, a drw, and some docs and writes those docs to the dolt_docs table
+// (perhaps creating it in the process). The table might not necessarily need to be created if there are no docs in the
+// repo yet.
+func UpdateRootWithDocs(ctx context.Context, dbData DbData, root *doltdb.RootValue, rootType RootType, docDetails doltdocs.Docs) (*doltdb.RootValue, error) {
```

Comment:
Wrong abstraction. You're doing too much here. We need two things:

1) methods that update root values with new tables or doc contents
2) methods that update the repo state of [working,staged,head] to root value hashes.

You're combining them here.

You don't need, and shouldn't have, methods that update repo state specific to docs contents. If you want to update docs persisted in the repo state, then

1) modify a root value to have the dolt_docs table you want,
2) write that thing to staged or head

This method should do 1) only, and it should go in the dolt_docs package.

Reply:
This method should take (ctx, root, docs) and return (root, error)